### PR TITLE
Compiler time evaluation of all int and bool operators.

### PR DESCRIPTION
**Pull request by csyonghe**
This PR implements compile time evaluation of all int and bool operators and the "?:" operator.

Fix code gen bug that causes `static const int` values defined in a struct type to emit as `integer_constant` field in target code.
The cause for this issue is that `static const` normally would lower as global insts, but if its value is dependent on a generic value parameter, then they will still be emitted inside the `IRStructType`. The solution is to introduce a hoisting pass to move these to global scope after IR specialization pass (when these constants no longer depend on any generic parameters).

Fixed a parser disambiguation issue when parsing things like `StructuredBuffer<MyType<1,2>>`.

Fixed a bug that causes COM interface definition not emitted in some situations. The fix makes sure that if an interface type is used from a `lookup_witness_table` inst, its definition will be emitted, even if it is accessed through a `NativePtr`. (By default, types referenced by `NativePtr` only need a forward declaration).

### CODE REVIEW

File: source/slang/slang-ast-type.cpp [hunk 1]
```diff
@@ -387,15 +387,19 @@ Val* ArrayExpressionType::_substituteImplOverride(ASTBuilder* astBuilder, Substi
...
-    SLANG_ASSERT(arrlen);
+    IntVal* newArrayLength = nullptr;
+    if (arrayLength)
+    {
+        newArrayLength = as<IntVal>(arrayLength->substituteImpl(astBuilder, subst, &diff));
+        SLANG_ASSERT(newArrayLength);
+    }
     if (diff)
     {
         *ioDiff = 1;
```
Comment (jsmall-nvidia) [inline source/slang/slang-ast-type.cpp:398]: Is there a reason to not increment, or add the diff?

File: source/slang/slang-ir-hoist-constants.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
+// slang-ir-hoist-constants.cpp
+#pragma once
```
Comment (jsmall-nvidia) [inline source/slang/slang-ir-hoist-constants.cpp:2]: #pragma once ?

**Review by jsmall-nvidia — APPROVED** [review]
Nice.

File: source/slang/slang-ast-type.cpp (hunk 1 above)
Comment (csyonghe) [reply source/slang/slang-ast-type.cpp:398]: Shouldn't matter. This is just used as a `bool`.

File: source/slang/slang-ir-hoist-constants.cpp (hunk 2 above)
Comment (csyonghe) [reply source/slang/slang-ir-hoist-constants.cpp]: Fixed.

**Pull request merged**